lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1444,6 +1444,68 @@ static void diagRecursivePropertyAccess(const Expr *E, const DeclContext *DC) {
   const_cast<Expr *>(E)->walk(walker);
 }
 
+static void
+diagnoseMemberMethodAssignmentsCapturingSelf(const Expr *E,
+                                             const DeclContext *DC) {
+
+  class DiagnoseWalker : public ASTWalker {
+    ASTContext &Ctx;
+
+  public:
+    explicit DiagnoseWalker(ASTContext &ctx) : Ctx(ctx) {}
+
+    bool walkToDeclPre(Decl *D) override { return false; }
+
+    bool shouldWalkIntoSeparatelyCheckedClosure(ClosureExpr *expr) override {
+      return false;
+    }
+
+    bool shouldWalkCaptureInitializerExpressions() override { return true; }
+
+    bool shouldWalkIntoTapExpression() override { return false; }
+
+    DeclRefExpr *getMemberFuncReferencingSelf(Expr *selfParam,
+                                              AutoClosureExpr *ACE) {
+
+      if (auto *ICE = dyn_cast<AutoClosureExpr>(ACE->getSingleExpressionBody()))
+        if (auto *AE = dyn_cast<ApplyExpr>(ICE->getSingleExpressionBody()))
+          if (auto *DSCE = dyn_cast<DotSyntaxCallExpr>(AE->getFn()))
+            if (auto *base = dyn_cast<DeclRefExpr>(DSCE->getBase()))
+              if (base->getType()->getCanonicalType() ==
+                  selfParam->getType()->getCanonicalType()) {
+                auto *member = dyn_cast<DeclRefExpr>(DSCE->getFn());
+                return member;
+              }
+
+      return nullptr;
+    }
+
+    std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+
+      if (auto *ASE = dyn_cast<AssignExpr>(E))
+        if (auto *AE = dyn_cast<ApplyExpr>(ASE->getSrc()))
```

Comment:
This pattern match across the AST is too domain-specific. For example, I can construct an AST that evades this matcher by surrounding the assignment source in parentheses.

Reply:
In general, the expression AST has the notion of "sugared" expressions that don't contribute meaningfully to the semantics of its sub-expressions. See `Expr::getSemanticsProvidingExpr()`

Reply:
Yes, makes sense. 

However, I went the route of folding the logic inside the current `diagnoseMemberMethodAssignmentsCapturingSelf`, which has changed things quite a bit. I am not checking for assignment expressions anymore and am also not sure about the logic's robustness. Sorry to make you review it once again :disappointed: